### PR TITLE
Umbraco 8 Update

### PR DESCRIPTION
**Pull request by marccameron3d**
Would love to talk it over with you if you get the time Matt, upgraded to allow for umbraco 8 using the new migration... technique?

I can only apply this PR to the `develop` branch but maybe good to have a v8 version unless the previous nuget builds are fine to live as they are.

Initially the framework needed updates to 4.7.2 to allow for the new UmbracoCMS.Core

**Updating Nugets:**
UmbracoCms.core version - 8.0+
Microsoft.AspNet.WebApi.WebHost - 5.2.7.0

After this, the next step was converting over how the code access's the database using the new `Composer.Current...`

Once this was complete you are able to run the `/oauth/token` endpoint successfully once you initialise the `OAuth.ConfigureEndpoint("/oauth/token", ...` method within a `IComponent` and register it using the `IUserComposer`.

This was the easy bit, although this was not able to use refresh token store or client store.

So I have removed the old migration runner and updated to use the new practice adopted by umbraco v8, which was good to learn. I have built 2 plans and included a helper to pull the state from the db. As it doesnt support Up / Down anymore I have included a remove as well as the initialise migrations.

I didnt expect to get where I was so happy to go over it and take some pointers from you about the code. Hope its useful for you, I will be using it actively in my current project and havent encountered any issues as of yet.

It could maybe do with a clean up or discussion over versioning the states as I know you used semver before, so it can either be removed now or states amended to use it.

### CODE REVIEW

**Comment by mattbrailsford** [comment]
Hey Marc,

This is awesome. I haven't looked through it all yet so I'll get back to you as soon as I can with feedback, but just wanted to say a big thank you in the meantime.

Matt

File: src/Our.Umbraco.AuthU/Component/MigrationsRunnerComponent.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+        public void Initialize()
+        {
+            try
+            {
+                //Create the oAuth tables
+                var upgrader = new Upgrader(new RegisterOAuthTables());
+
+                //Run this plan to remove the oAuth tables, there may be a better way to do this?
+                //var upgrader = new Upgrader(new RemoveOAuthTables());
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Component/MigrationsRunnerComponent.cs]: To be honest, I don't think we need to worry about downgrading. Umbraco have removed the ability to undo a migration, so I'd say lets just focus on installing.

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Remove/DeleteAuthClientStoreTable.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using Our.Umbraco.AuthU.Models;
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Remove/DeleteAuthClientStoreTable.cs]: Doing away with the downgrader, this won't be needed

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Add/CreateAuthClientStoreTable.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿using Our.Umbraco.AuthU.Models;
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Add/CreateAuthClientStoreTable.cs:1]: These files don't need to be an an Add folder.

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Add/CreateDemoClient.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Our.Umbraco.AuthU.Data.Migrations.UmbracoDbOAuthClientStore
+{
+    /// <summary>
+    /// Simple migration to add a demo client to the OAuthClient Table
+    /// </summary>
+    internal class CreateDemoClient : MigrationBase
+	{
+		public CreateDemoClient(IMigrationContext context) : base(context)
+		{ }
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Add/CreateDemoClient.cs]: Formatting appears to be a bit off here. Double check tabs vs spaces.

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Add/CreateDemoClient.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+        public override void Migrate()
+        {
+            var existing = Database.SingleOrDefault<OAuthClient>($"SELECT * FROM [OAuthClient] WHERE [ClientId] = @0", "DemoClient");
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Add/CreateDemoClient.cs]: I'd probably change this query to `SELECT 1 FROM` as we don't actually need any data, we are just checking if anything exists. Fetching all the data from the DB is a waste here so lets make this as slim as we can.

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Remove/DeleteAuthClientStoreTable.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+namespace Our.Umbraco.AuthU.Data.Migrations.UmbracoDbOAuthClientStore
+{
+    /// <summary>
+    /// Simple migration to delete the OAuthClient Table
+    /// </summary>
+    internal class DeleteAuthClientStoreTable : MigrationBase
+	{
+		public DeleteAuthClientStoreTable(IMigrationContext context) 
+			: base(context)
+		{ }
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthClientStore/Remove/DeleteAuthClientStoreTable.cs]: Formatting

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthRefreshTokenStore/Add/AddDeviceIdColumn.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Our.Umbraco.AuthU.Data.Migrations.UmbracoDbOAuthRefreshTokenStore
+{
+    /// <summary>
+    /// Simple migration to create the DeviceId column on the OAuthRefreshToken Table
+    /// </summary>
+    internal class AddDeviceIdColumn : MigrationBase
+	{
+		public AddDeviceIdColumn(IMigrationContext context) : base(context)
+		{ }
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthRefreshTokenStore/Add/AddDeviceIdColumn.cs]: Formatting

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthRefreshTokenStore/Add/AddDeviceIdColumn.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+        public override void Migrate()
+        {
+            if (!ColumnExists("OAuthRefreshToken", "DeviceId"))
+            {
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthRefreshTokenStore/Add/AddDeviceIdColumn.cs]: For consistency, we should either do away with the curly brackets, or make sure we use curly brackets everywhere. My usual rule is, if it's a one line statement, then don't use them.

File: src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthRefreshTokenStore/Remove/DeleteAuthRefreshTokenStoreTable.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿using Our.Umbraco.AuthU.Models;
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Migrations/UmbracoDbOAuthRefreshTokenStore/Remove/DeleteAuthRefreshTokenStoreTable.cs]: Not needed

File: src/Our.Umbraco.AuthU/Data/Plans/RegisterOAuthTables.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                .To<CreateAuthRefreshTokenStoreTable>("CreateAuthRefreshTokenStoreTable")
+                .To<AddDeviceIdColumn>("AddDeviceIdColumn")
+                .To<CreateDemoClient>("CreateDemoClient");
+            }
+        }
+
+        /// <summary>
+        /// Using the helper methid to get the AuthU Key 'Umbraco.Core.Upgrader.State+AuthU' Value
+        /// </summary>
+        public override string InitialState => AuthUKeyValueHelper.GetValue();
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Plans/RegisterOAuthTables.cs]: I don't think this is quite right. I don't think you want to be looking up the current state from the DB, rather I think this is more for when you are upgrading from like v7. Given our install scripts are defensive, only applying when needed, I think we might be able to just have this as an empty string.

File: src/Our.Umbraco.AuthU/Data/Plans/RemoveOAuthTables.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
+﻿using Our.Umbraco.AuthU.Data.Helpers;
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Plans/RemoveOAuthTables.cs]: Not needed

File: src/Our.Umbraco.AuthU/Data/UmbracoDbOAuthClientStore.cs [hunk 12]
```diff
@@ -1,27 +1,16 @@
...
-		internal const string CurrentVersion = "1.0.0";
-		internal const string SubProductName = "AuthU_UmbracoDbOAuthClientStore";
-
-		protected Database Db => ApplicationContext.Current.DatabaseContext.Database;
-
-        public UmbracoDbOAuthClientStore()
-        {
-			MigrationsRunner.RunMigrations(CurrentVersion, SubProductName);
-		}
+		protected Database Db => new Database("umbracoDbDsn");
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/UmbracoDbOAuthClientStore.cs]: Should be reusing the DB from the umbraco database context like it was before.

File: src/Our.Umbraco.AuthU/Data/UmbracoDbOAuthRefreshTokenStore.cs [hunk 13]
```diff
@@ -11,12 +9,7 @@ public class UmbracoDbOAuthRefreshTokenStore : IOAuthRefreshTokenStore
 		internal const string CurrentVersion = "1.0.1";
 		internal const string SubProductName = "AuthU_UmbracoDbOAuthRefreshTokenStore";
 
-		protected Database Db => ApplicationContext.Current.DatabaseContext.Database;
-
-        public UmbracoDbOAuthRefreshTokenStore()
-        {
-			MigrationsRunner.RunMigrations(CurrentVersion, SubProductName);
-		}
+        protected Database Db => new Database("umbracoDbDsn");
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/UmbracoDbOAuthRefreshTokenStore.cs]: Reuse umbraco database

File: src/Our.Umbraco.AuthU/Extensions/StringExtensions.cs [hunk 14]
```diff
@@ -6,7 +6,8 @@ namespace Our.Umbraco.AuthU.Extensions
 {
     public static class StringExtensions
     {
-        internal static string GenerateHash(this string input)
+        //renamed as it was conflicting with Umbraco.Core.StringHelpers
+        internal static string oAuthGenerateHash(this string input)
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Extensions/StringExtensions.cs]: I'd probably rename this be be more natural, ie `GenerateOAuthHash`

File: src/Our.Umbraco.AuthU/Data/Plans/RegisterOAuthTables.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        /// <summary>
+        /// Add the demo id if preferred
+        /// </summary>
+        /// <param name="CreateDemoClient"></param>
+        public RegisterOAuthTables(bool CreateDemoClient = false) : base("AuthU")
+        {
+
+            if (!CreateDemoClient)
+            {
+                From(InitialState)
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Plans/RegisterOAuthTables.cs]: Duplication here, but given the above comment, I think we only need one route anyway.

File: src/Our.Umbraco.AuthU/Data/Plans/RegisterOAuthTables.cs [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    /// <summary>
+    /// This is the plan that runs the AuthU Migrations
+    /// </summary>
+    internal class RegisterOAuthTables : MigrationPlan
+    {
+        /// <summary>
+        /// Add the demo id if preferred
+        /// </summary>
+        /// <param name="CreateDemoClient"></param>
+        public RegisterOAuthTables(bool CreateDemoClient = false) : base("AuthU")
```
Comment (mattbrailsford) [inline src/Our.Umbraco.AuthU/Data/Plans/RegisterOAuthTables.cs]: `CreateDemoClient` is never true. I think we could probably just remove this param and just always install demo client. If people don't use it, it's not a big deal.

**Review by mattbrailsford — CHANGES_REQUESTED** [review]
Hey Marc, sorry for the delay on this.

It's a really good first stab at an update, so thanks for putting in the time to do this. I've just raised a few things to look into. Feel free to ping me if you want to go through anything.

**Comment by marccameron3d** [comment]
Ill change the DB context to use IScopeProvider, unless you know of a better way? I think registering this as a difference service may be good so we can use it in the AuthClientStore and RefreshTokenStore without the need to add the param to the constructor of each when calling them?

**Review by marccameron3d — COMMENTED** [review]
I have updated the comments you mentioned, my concern is with the database context and if using Current is the best way?

**Comment by marccameron3d** [comment]
@qz2rg4 Hey mate, the changes I have implemented on this PR work for me on a prod project. There is one ailment I found relating to header being set twice for CORS but I can add that issue in. The changes are failing to build but @mattbrailsford will review when he has time. In the meanwhile happy to help you out if you need it. @marcD8 is an old account from a previous role so hit me up on @marccameron3d

**Pull request merged**